lifetimes do not match method in trait

When using `async-trait` the compiler reports that the lifetimes do not match when no lifetimes are specified.  Specifying a lifetime resolves the issue even though the lifetime is never used.  

This crate requires `#![feature(proc_macro_hygiene)]` and thus requires **nightly**.  
```
$ cargo check                                                                                                                                                                                                               master
    Checking enzyme v0.1.0 (/usr/dataplus/enzyme)
    Checking user v0.1.0 (/usr/dataplus/enzyme/example/microservices/user)
error[E0195]: lifetime parameters or bounds on method `from_parts` do not match the trait declaration
  --> src/context.rs:10:24
   |
10 |     async fn from_parts(_parts: Parts, _params: Params) -> WebResult<TokenContext> {
   |                        ^ lifetimes do not match method in trait

error[E0195]: lifetime parameters or bounds on method `from_parts` do not match the trait declaration
  --> src/context.rs:21:24
   |
21 |     async fn from_parts(parts: Parts, _params: Params) -> WebResult<AuthContext> {
   |                        ^ lifetimes do not match method in trait

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0195`.
error: could not compile `user`.

To learn more, run the command again with --verbose.

```
### Trait Definition:
https://github.com/technetos/enzyme/blob/master/src/context.rs#L7

### Trait Usage:
https://github.com/technetos/enzyme/blob/master/example/microservices/user/src/context.rs#L10
and
https://github.com/technetos/enzyme/blob/master/example/microservices/user/src/context.rs#L21

### Method of reproduction:
Remove the `<'a>` lifetime from the trait definition and the usages above.  Run `cargo check`

--- a/src/expand.rs
+++ b/src/expand.rs
@@ -133,73 +133,63 @@ fn transform_sig(
         }
     }
 
-    let lifetime: Lifetime;
-    if !sig.generics.params.is_empty()
-        || !lifetimes.elided.is_empty()
-        || !lifetimes.explicit.is_empty()
-        || has_self
+    let where_clause = sig
+        .generics
+        .where_clause
+        .get_or_insert_with(|| WhereClause {
+            where_token: Default::default(),
+            predicates: Punctuated::new(),
+        });
+    for param in sig
+        .generics
+        .params
+        .iter()
+        .chain(context.lifetimes(&lifetimes.explicit))
     {
-        lifetime = parse_quote!('async_trait);
-        let where_clause = sig
-            .generics
-            .where_clause
-            .get_or_insert_with(|| WhereClause {
-                where_token: Default::default(),
-                predicates: Punctuated::new(),
-            });
-        for param in sig
-            .generics
-            .params
-            .iter()
-            .chain(context.lifetimes(&lifetimes.explicit))
-        {
-            match param {
-                GenericParam::Type(param) => {
-                    let param = &param.ident;
-                    where_clause
-                        .predicates
-                        .push(parse_quote!(#param: #lifetime));
-                }
-                GenericParam::Lifetime(param) => {
-                    let param = &param.lifetime;
-                    where_clause
-                        .predicates
-                        .push(parse_quote!(#param: #lifetime));
-                }
-                GenericParam::Const(_) => {}
+        match param {
+            GenericParam::Type(param) => {
+                let param = &param.ident;
+                where_clause
+                    .predicates
+                    .push(parse_quote!(#param: 'async_trait));
             }
+            GenericParam::Lifetime(param) => {
+                let param = &param.lifetime;
+                where_clause
+                    .predicates
+                    .push(parse_quote!(#param: 'async_trait));
+            }
+            GenericParam::Const(_) => {}
         }
-        for elided in lifetimes.elided {
-            sig.generics.params.push(parse_quote!(#elided));
-            where_clause
-                .predicates
-                .push(parse_quote!(#elided: #lifetime));
-        }
-        sig.generics.params.push(parse_quote!(#lifetime));
-        if has_self {
-            let bound: Ident = match sig.inputs.iter().next() {
-                Some(FnArg::Receiver(Receiver {
-                    reference: Some(_),
-                    mutability: None,
-                    ..
-                })) => parse_quote!(Sync),
-                _ => parse_quote!(Send),
-            };
-            let assume_bound = match context {
-                Context::Trait { supertraits, .. } => {
-                    !has_default || has_bound(supertraits, &bound)
-                }
-                Context::Impl { .. } => true,
-            };
-            where_clause.predicates.push(if assume_bound || is_local {
-                parse_quote!(Self: #lifetime)
-            } else {
-                parse_quote!(Self: ::core::marker::#bound + #lifetime)
-            });
-        }
-    } else {
-        lifetime = parse_quote!('static);
-    };
+    }
+    for elided in lifetimes.elided {
+        sig.generics.params.push(parse_quote!(#elided));
+        where_clause
+            .predicates
+            .push(parse_quote!(#elided: 'async_trait));
+    }
+    sig.generics.params.push(parse_quote!('async_trait));
+    if has_self {
+        let bound: Ident = match sig.inputs.iter().next() {
+            Some(FnArg::Receiver(Receiver {
+                reference: Some(_),
+                mutability: None,
+                ..
+            })) => parse_quote!(Sync),
+            _ => parse_quote!(Send),
+        };
+        let assume_bound = match context {
+            Context::Trait { supertraits, .. } => {
+                !has_default || has_bound(supertraits, &bound)
+            }
+            Context::Impl { .. } => true,
+        };
+        where_clause.predicates.push(if assume_bound || is_local {
+            parse_quote!(Self: 'async_trait)
+        } else {
+            parse_quote!(Self: ::core::marker::#bound + 'async_trait)
+        });
+    }
 
     for (i, arg) in sig.inputs.iter_mut().enumerate() {
         match arg {
@@ -220,9 +210,9 @@ fn transform_sig(
     }
 
     let bounds = if is_local {
-        quote!(#lifetime)
+        quote!('async_trait)
     } else {
-        quote!(::core::marker::Send + #lifetime)
+        quote!(::core::marker::Send + 'async_trait)
     };
 
     sig.output = parse_quote! {

--- a/tests/test.rs
+++ b/tests/test.rs
@@ -375,3 +375,22 @@ pub mod issue31 {
         }
     }
 }
+
+// https://github.com/dtolnay/async-trait/issues/42
+pub mod issue42 {
+    use async_trait::async_trait;
+
+    #[async_trait]
+    pub trait Context: Sized {
+        async fn from_parts() -> Self;
+    }
+
+    pub struct TokenContext;
+
+    #[async_trait]
+    impl Context for TokenContext {
+        async fn from_parts() -> TokenContext {
+            TokenContext
+        }
+    }
+}

